File: 86. Partition List/main.c
#include <stdio.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *partition(struct ListNode *head, int x) {
    struct ListNode *newHead = NULL;
    struct ListNode **first = &newHead;
    struct ListNode **second = &head;

    while (*second != NULL) {
        if ((*second)->val < x) {
            *first = *second;
            *second = (*second)->next;
            first = &(*first)->next;
        } else {
            second = &(*second)->next;
        }
    }

    *first = head;

    return newHead;
}

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *partition(struct ListNode *head, int x) {
    struct ListNode dummy1;
    struct ListNode dummy2;
    struct ListNode *first = &dummy1;
    struct ListNode *second = &dummy2;

    while (head != NULL) {
        if (head->val < x) {
            first->next = head;
            first = first->next;
        } else {
            second->next = head;
            second = second->next;
        }
        head = head->next;
    }

    second->next = NULL;
    first->next = dummy2.next;

    return dummy1.next;
}

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *partition(struct ListNode *head, int x) {
    struct ListNode dummy1;
    struct ListNode dummy2;
    struct ListNode *first = &dummy1;
    struct ListNode *second = &dummy2;
    dummy1.next = head;

    while (first->next != NULL) {
        if (first->next->val < x) {
            first = first->next;
        } else {
            second->next = first->next;
            first->next = first->next->next;
            second = second->next;
        }
    }

    second->next = NULL;
    first->next = dummy2.next;

    return dummy1.next;
}

File: 86. Partition List/main.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        dummy1 = ListNode()
        dummy2 = ListNode()
        first = dummy1
        second = dummy2

        while head:
            if head.val < x:
                first.next = head
                first = first.next
            else:
                second.next = head
                second = second.next

            head = head.next

        first.next = dummy2.next
        second.next = None

        return dummy1.next


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        dummy1 = ListNode(next=head)
        dummy2 = ListNode()
        first = dummy1
        second = dummy2

        while first.next:
            if first.next.val < x:
                first = first.next
            else:
                second.next = first.next
                second = second.next
                first.next = first.next.next

            head = head.next

        first.next = dummy2.next
        second.next = None

        return dummy1.next
